lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -141,93 +146,174 @@
-    for package in apt_installed_packages:
-        ac_pkg = apt_cache[package].installed
-        package_details = dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, source='apt')
-        if package_details['name'] not in installed_packages:
-            installed_packages[package_details['name']] = [package_details]
-        else:
-            installed_packages[package_details['name']].append(package_details)
-    return installed_packages
-
-
-# FIXME: add more listing methods
+class PkgMgr(object):
+
+    def __init__(self, module):
+
+        self.m = module
+
+    @abstractmethod
+    def test(self):
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        pass
+
+    @abstractmethod
+    def get_package_details(self):
+        pass
+
+    def get_packages(self):
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.name
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
+
+    def __init__(self, module):
+
+        super(LibMgr, self).__init__(module)
+        self.lib = None
+
+    def test(self):
+        found = False
+        try:
+            self.lib = __import__(self.LIB)
+            found = True
+        except ImportError:
+            pass
+        return found
+
+
+class CLIMgr(PkgMgr):
+
+    CLI = None
+
+    def __init__(self, module):
+
+        super(CLIMgr, self).__init__(module)
+        self.cli = None
+
+    def test(self):
+        self.cli = self.m.get_bin_path(self.CLI, False)
+        return bool(self.cli)
+
+
+class RPM(LibMgr):
+
+    name = 'rpm'
+    LIB = 'rpm'
+
+    def list_installed(self):
+        return self.lib.TransactionSet().dbMatch()
+
+    def get_package_details(self, package):
+        return dict(name=package[self.lib.RPMTAG_NAME],
+                    version=package[self.lib.RPMTAG_VERSION],
+                    release=package[self.lib.RPMTAG_RELEASE],
+                    epoch=package[self.lib.RPMTAG_EPOCH],
+                    arch=package[self.lib.RPMTAG_ARCH],)
+
+
+class APT(LibMgr):
+
+    name = 'apt'
+    LIB = 'apt'
+
+    def test(self):
+        works = super(APT, self).test()
+        if works:
+            self._cache = self.lib.Cache()
+        return works
+
+    def list_installed(self):
+        return [pk for pk in self._cache.keys() if self._cache[pk].is_installed]
+
+    def get_package_details(self, package):
+        ac_pkg = self._cache[package].installed
+        return dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture)
+
+
+class PORTAGE(CLIMgr):
+
+    name = 'portage'
+    CLI = 'qlist'
```

Comment:
Can we use `eix` when it's present?

Reply:
yes and no, eix-installed gives similar info but not the same, i can make a conditional branch but the results will differ.